File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isLower, isSpace)

data Tone = Normal | Loud | Unknown

data Expression = Silence | Statement | Question

responseFor :: String -> String
responseFor xs = respond (readInput xs)

respond :: (Tone, Expression) -> String
respond (Loud, _)     = "Whoa, chill out!"
respond (_, Question) = "Sure."
respond (_, Silence)  = "Fine. Be that way!"
respond (_, _)        = "Whatever."

readInput :: String -> (Tone, Expression)
readInput = readNextChar (Unknown, Silence)

readNextChar :: (Tone, Expression) -> String -> (Tone, Expression)
readNextChar (t,e) []       = (t,e)
readNextChar (t,_) ('?':xs) = readNextChar (t, Question) xs
readNextChar (Unknown, Silence) (x:xs)
    | isSpace x             = readNextChar (Unknown, Silence) xs
    | isLower x             = readNextChar (Normal, Statement) xs
    | isUpper x             = readNextChar (Loud, Statement) xs
    | otherwise             = readNextChar (Unknown, Statement) xs
readNextChar (Unknown, exp) (x:xs)
    | isSpace x             = readNextChar (Unknown, exp) xs
    | isLower x             = readNextChar (Normal, Statement) xs
    | isUpper x             = readNextChar (Loud, Statement) xs
    | otherwise             = readNextChar (Unknown, Statement) xs
readNextChar (Loud, exp) (x:xs)
    | isSpace x             = readNextChar (Loud, exp) xs
    | isLower x             = readNextChar (Normal, Statement) xs
    | otherwise             = readNextChar (Loud, Statement) xs
readNextChar (Normal, exp) (x:xs)
    | isSpace x             = readNextChar (Normal, exp) xs
    | otherwise             = readNextChar (Normal, Statement) xs

File: go/bob/bob.go
// Package bob implements an annoying teenager
package bob

import (
	"regexp"
	"strings"
)

var alphaRE = regexp.MustCompile(`[a-zA-Z]`)
var lowerRE = regexp.MustCompile(`[a-z]`)

// Hey implements an annoying teenager
func Hey(remark string) string {

	remark = strings.TrimSpace(remark)

	if len(remark) == 0 {
		return "Fine. Be that way!"
	}

	question := strings.HasSuffix(remark, "?")
	shouting := alphaRE.MatchString(remark) && !lowerRE.MatchString(remark)

	if question && shouting {
		return "Calm down, I know what I'm doing!"
	} else if question {
		return "Sure."
	} else if shouting {
		return "Whoa, chill out!"
	}

	return "Whatever."
}
